File: packages/schemas/src/command-room-worker.ts
import { z } from "zod";

const boundedId = z.string().min(1).max(256);

export const GatewayChannelSchema = z.enum(["command_room", "agent_runs"]);

export const GatewaySessionSourceSchema = z.object({
  channel: GatewayChannelSchema,
  conversationId: boundedId,
  projectId: boundedId.optional(),
  threadId: boundedId.optional(),
  workspaceId: boundedId.optional(),
  snapshotId: boundedId.optional(),
  documentRevisionId: boundedId.optional(),
}).strict();

export const GatewayCommandRoomSessionSourceSchema = GatewaySessionSourceSchema.extend({
  channel: z.literal("command_room"),
});

export const GatewayTurnMessageSchema = z.object({
  role: z.enum(["user", "assistant"]),
  content: z.string().min(1).max(32_000),
}).strict();

export const GatewayReasoningEffortSchema = z.enum(["low", "medium", "high", "max"]);
export const GatewayThinkingSchema = z.enum(["on", "off"]);
export const GatewayRequestStyleSchema = z.enum(["chat-completions", "responses"]);

export const GatewayTurnRequestSchema = z.object({
  mode: z.literal("work"),
  runId: boundedId.optional(),
  session: GatewayCommandRoomSessionSourceSchema,
  messages: z.array(GatewayTurnMessageSchema).min(1).max(40),
  modelAlias: z.string().min(1).max(64),
  reasoningEffort: GatewayReasoningEffortSchema.default("high"),
  thinking: GatewayThinkingSchema.default("on"),
  clientCorrelationId: z.string().regex(/^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/).optional(),
}).strict();

export const GatewayBindingSchema = z.object({
  channel: z.string().min(1),
  tenantId: z.string().min(1),
  actorId: z.string().min(1),
  conversationId: z.string().min(1),
  projectId: z.string().min(1).optional(),
  threadId: z.string().min(1).optional(),
  workspaceId: z.string().min(1).optional(),
  snapshotId: z.string().min(1).optional(),
  documentRevisionId: z.string().min(1).optional(),
}).strict();

export const GatewayModelProfileSchema = z.object({
  alias: z.string().min(1),
  provider: z.string().min(1),
  model: z.string().min(1),
  transport: z.enum(["openai-compatible", "native"]),
  requestStyle: GatewayRequestStyleSchema.default("chat-completions"),
  supportsThinking: z.boolean(),
  selectedEffort: GatewayReasoningEffortSchema,
  thinking: GatewayThinkingSchema,
  reasoningEffortMap: z.record(z.string()).optional(),
}).strict();

export const GatewayPromptSectionSizesSchema = z.object({
  stable: z.number().int().nonnegative(),
  context: z.number().int().nonnegative(),
  volatile: z.number().int().nonnegative(),
}).strict();

export const GatewayPromptMetadataSchema = z.object({
  version: z.string().min(1),
  stableHash: z.string().regex(/^[0-9a-fA-F]{64}$/),
  sectionSizes: GatewayPromptSectionSizesSchema,
  injectionFindings: z.array(z.string()),
}).strict();

export const GatewayTurnPreparedSchema = z.object({
  protocolVersion: z.literal("command-room.gateway.v1"),
  runId: z.string().min(1),
  sessionId: z.string().min(1),
  sessionKeyFingerprint: z.string().regex(/^[0-9a-fA-F]{64}$/),
  binding: GatewayBindingSchema,
  profile: GatewayModelProfileSchema,
  prompt: GatewayPromptMetadataSchema,
  handoff: z.enum(["service-conversation-loop", "legacy-web-provider"]),
}).strict();

export const GatewayWorkEventTypeSchema = z.enum([
  "turn.started",
  "status.update",
  "assistant.interim",
  "reasoning.delta",
  "plan.updated",
  "tool.generating",
  "tool.started",
  "tool.progress",
  "tool.completed",
  "tool.output_risk",
  "approval.requested",
  "approval.resolved",
  "subagent.started",
  "subagent.progress",
  "subagent.completed",
  "background.completed",
  "artifact.created",
  "log.line",
  "assistant.delta",
  "turn.completed",
  "error",
]);

const WorkTaskSchema = z.object({
  id: boundedId,
  title: boundedId,
  state: z.enum(["pending", "in_progress", "completed", "failed", "cancelled"]),
  detail: z.string().max(16_000).optional(),
  startedAt: z.string().min(1).optional(),
  completedAt: z.string().min(1).optional(),
}).strict();

const WorkToolSchema = z.object({
  toolId: boundedId,
  name: boundedId,
  state: z.enum(["generating", "running", "completed", "failed"]),
  args: z.unknown().optional(),
  result: z.unknown().optional(),
  summary: z.string().max(16_000).optional(),
  progress: z.string().max(16_000).optional(),
  durationMs: z.number().int().nonnegative().optional(),
  startedAt: z.string().min(1).optional(),
  completedAt: z.string().min(1).optional(),
}).strict();

const WorkApprovalSchema = z.object({
  approvalId: boundedId,
  action: boundedId,
  reason: z.string().min(1).max(16_000),
  args: z.unknown().optional(),
  createdAt: z.string().min(1),
  expiresAt: z.string().min(1),
  state: z.enum(["pending", "approved", "denied", "expired"]),
}).strict();

const WorkArtifactSchema = z.object({
  artifactId: boundedId,
  name: boundedId,
  kind: z.string().max(256).optional(),
  uri: z.string().max(2_048).optional(),
  sizeBytes: z.number().int().nonnegative().optional(),
  summary: z.string().max(16_000).optional(),
  createdAt: z.string().min(1).optional(),
}).strict();

const WorkLogSchema = z.object({
  level: z.enum(["debug", "info", "warn", "error"]),
  text: z.string().min(1).max(16_000),
}).strict();

const workEventCommon = {
  runId: boundedId,
  conversationId: boundedId,
  eventId: boundedId,
  sequence: z.number().int().nonnegative(),
  timestamp: z.string().min(1),
} as const;

function workEvent<TType extends z.ZodLiteral<string>, TShape extends z.ZodRawShape>(type: TType, shape: TShape) {
  return z.object({ ...workEventCommon, type, ...shape }).strict();
}

const WorkEventTurnStartedSchema = workEvent(z.literal("turn.started"), {
  phase: z.string().max(256).optional(),
  statusLabel: z.string().max(256).optional(),
  statusDetail: z.string().max(16_000).optional(),
});
const WorkEventStatusSchema = workEvent(z.literal("status.update"), {
  phase: z.string().max(256).optional(),
  statusLabel: z.string().max(256).optional(),
  statusDetail: z.string().max(16_000).optional(),
});
const WorkEventInterimSchema = workEvent(z.literal("assistant.interim"), { message: z.string().max(16_000) });
const WorkEventReasoningSchema = workEvent(z.literal("reasoning.delta"), { delta: z.string().max(16_000) });
const WorkEventPlanSchema = workEvent(z.literal("plan.updated"), { tasks: z.array(WorkTaskSchema).max(256) });
const WorkEventToolGeneratingSchema = workEvent(z.literal("tool.generating"), { tool: WorkToolSchema });
const WorkEventToolStartedSchema = workEvent(z.literal("tool.started"), { tool: WorkToolSchema });
const WorkEventToolProgressSchema = workEvent(z.literal("tool.progress"), {
  tool: WorkToolSchema.optional(),
  progress: z.string().max(16_000).optional(),
});
const WorkEventToolCompletedSchema = workEvent(z.literal("tool.completed"), { tool: WorkToolSchema });
const WorkEventToolRiskSchema = workEvent(z.literal("tool.output_risk"), {
  tool: WorkToolSchema,
  summary: z.string().max(16_000).optional(),
});
const WorkEventApprovalRequestedSchema = workEvent(z.literal("approval.requested"), { approval: WorkApprovalSchema });
const WorkEventApprovalResolvedSchema = workEvent(z.literal("approval.resolved"), { approval: WorkApprovalSchema });
const WorkEventSubagentStartedSchema = workEvent(z.literal("subagent.started"), { summary: z.string().max(16_000).optional() });
const WorkEventSubagentProgressSchema = workEvent(z.literal("subagent.progress"), { progress: z.string().max(16_000).optional() });
const WorkEventSubagentCompletedSchema = workEvent(z.literal("subagent.completed"), { summary: z.string().max(16_000).optional() });
const WorkEventBackgroundSchema = workEvent(z.literal("background.completed"), { summary: z.string().max(16_000).optional() });
const WorkEventArtifactSchema = workEvent(z.literal("artifact.created"), { artifact: WorkArtifactSchema });
const WorkEventLogSchema = workEvent(z.literal("log.line"), { log: WorkLogSchema });
const WorkEventAssistantDeltaSchema = workEvent(z.literal("assistant.delta"), { delta: z.string().max(16_000) });
const WorkEventCompletedSchema = workEvent(z.literal("turn.completed"), {
  finalMarkdown: z.string().max(32_000).optional(),
  stopReason: z.string().max(256).optional(),
});
const WorkEventErrorSchema = workEvent(z.literal("error"), {
  errorCode: z.string().max(256).optional(),
  errorMessage: z.string().max(16_000).optional(),
  retryable: z.boolean().optional(),
});

export const GatewayWorkEventSchema = z.discriminatedUnion("type", [
  WorkEventTurnStartedSchema,
  WorkEventStatusSchema,
  WorkEventInterimSchema,
  WorkEventReasoningSchema,
  WorkEventPlanSchema,
  WorkEventToolGeneratingSchema,
  WorkEventToolStartedSchema,
  WorkEventToolProgressSchema,
  WorkEventToolCompletedSchema,
  WorkEventToolRiskSchema,
  WorkEventApprovalRequestedSchema,
  WorkEventApprovalResolvedSchema,
  WorkEventSubagentStartedSchema,
  WorkEventSubagentProgressSchema,
  WorkEventSubagentCompletedSchema,
  WorkEventBackgroundSchema,
  WorkEventArtifactSchema,
  WorkEventLogSchema,
  WorkEventAssistantDeltaSchema,
  WorkEventCompletedSchema,
  WorkEventErrorSchema,
]);

export type GatewayChannel = z.infer<typeof GatewayChannelSchema>;
export type GatewaySessionSource = z.infer<typeof GatewaySessionSourceSchema>;
export type GatewayCommandRoomSessionSource = z.infer<typeof GatewayCommandRoomSessionSourceSchema>;
export type GatewayTurnMessage = z.infer<typeof GatewayTurnMessageSchema>;
export type GatewayTurnRequest = z.infer<typeof GatewayTurnRequestSchema>;
export type GatewayBinding = z.infer<typeof GatewayBindingSchema>;
export type GatewayModelProfile = z.infer<typeof GatewayModelProfileSchema>;
export type GatewayRequestStyle = z.infer<typeof GatewayRequestStyleSchema>;
export type GatewayPromptSectionSizes = z.infer<typeof GatewayPromptSectionSizesSchema>;
export type GatewayPromptMetadata = z.infer<typeof GatewayPromptMetadataSchema>;
export type GatewayTurnPrepared = z.infer<typeof GatewayTurnPreparedSchema>;
export type GatewayWorkEventType = z.infer<typeof GatewayWorkEventTypeSchema>;
export type GatewayWorkEvent = z.infer<typeof GatewayWorkEventSchema>;

File: packages/schemas/python/paax_schemas/command_room_worker.py
"""Pydantic mirror of the Command Room worker gateway wire contract."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictInt, StrictStr, model_validator

GatewayChannel = Literal["command_room", "agent_runs"]
GatewayRequestStyle = Literal["chat-completions", "responses"]


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    @model_validator(mode="before")
    @classmethod
    def reject_explicit_nulls(cls, value: Any) -> Any:
        """Mirror Zod optional fields: omitted is valid, explicit null is not."""
        if isinstance(value, dict) and any(item is None for item in value.values()):
            raise ValueError("null is not allowed for this wire contract")
        return value


class GatewaySessionSource(_StrictModel):
    channel: GatewayChannel
    conversationId: StrictStr = Field(min_length=1, max_length=256)
    projectId: StrictStr | None = Field(default=None, min_length=1, max_length=256)
    threadId: StrictStr | None = Field(default=None, min_length=1, max_length=256)
    workspaceId: StrictStr | None = Field(default=None, min_length=1, max_length=256)
    snapshotId: StrictStr | None = Field(default=None, min_length=1, max_length=256)
    documentRevisionId: StrictStr | None = Field(default=None, min_length=1, max_length=256)


class GatewayCommandRoomSessionSource(GatewaySessionSource):
    channel: Literal["command_room"]


class GatewayTurnMessage(_StrictModel):
    role: Literal["user", "assistant"]
    content: StrictStr = Field(min_length=1, max_length=32_000)


class GatewayTurnRequest(_StrictModel):
    mode: Literal["work"]
    runId: StrictStr | None = Field(default=None, min_length=1, max_length=256)
    session: GatewayCommandRoomSessionSource
    messages: list[GatewayTurnMessage] = Field(min_length=1, max_length=40)
    modelAlias: StrictStr = Field(min_length=1, max_length=64)
    reasoningEffort: Literal["low", "medium", "high", "max"] = "high"
    thinking: Literal["on", "off"] = "on"
    clientCorrelationId: StrictStr | None = Field(
        default=None,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$",
    )


class GatewayBinding(_StrictModel):
    channel: StrictStr = Field(min_length=1)
    tenantId: StrictStr = Field(min_length=1)
    actorId: StrictStr = Field(min_length=1)
    conversationId: StrictStr = Field(min_length=1)
    projectId: StrictStr | None = Field(default=None, min_length=1)
    threadId: StrictStr | None = Field(default=None, min_length=1)
    workspaceId: StrictStr | None = Field(default=None, min_length=1)
    snapshotId: StrictStr | None = Field(default=None, min_length=1)
    documentRevisionId: StrictStr | None = Field(default=None, min_length=1)


class GatewayModelProfile(_StrictModel):
    alias: StrictStr = Field(min_length=1)
    provider: StrictStr = Field(min_length=1)
    model: StrictStr = Field(min_length=1)
    transport: Literal["openai-compatible", "native"]
    requestStyle: GatewayRequestStyle = "chat-completions"
    supportsThinking: StrictBool
    selectedEffort: Literal["low", "medium", "high", "max"]
    thinking: Literal["on", "off"]
    reasoningEffortMap: dict[StrictStr, StrictStr] | None = None


class GatewayPromptSectionSizes(_StrictModel):
    stable: StrictInt = Field(ge=0)
    context: StrictInt = Field(ge=0)
    volatile: StrictInt = Field(ge=0)


class GatewayPromptMetadata(_StrictModel):
    version: StrictStr = Field(min_length=1)
    stableHash: StrictStr = Field(min_length=64, max_length=64, pattern=r"^[0-9a-fA-F]{64}$")
    sectionSizes: GatewayPromptSectionSizes
    injectionFindings: list[StrictStr]


class GatewayTurnPrepared(_StrictModel):
    protocolVersion: Literal["command-room.gateway.v1"]
    runId: StrictStr = Field(min_length=1)
    sessionId: StrictStr = Field(min_length=1)
    sessionKeyFingerprint: StrictStr = Field(min_length=64, max_length=64, pattern=r"^[0-9a-fA-F]{64}$")
    binding: GatewayBinding
    profile: GatewayModelProfile
    prompt: GatewayPromptMetadata
    handoff: Literal["service-conversation-loop", "legacy-web-provider"]


GatewayWorkEventType = Literal[
    "turn.started", "status.update", "assistant.interim", "reasoning.delta", "plan.updated",
    "tool.generating", "tool.started", "tool.progress", "tool.completed", "tool.output_risk",
    "approval.requested", "approval.resolved", "subagent.started", "subagent.progress",
    "subagent.completed", "background.completed", "artifact.created", "log.line",
    "assistant.delta", "turn.completed", "error",
]


class GatewayWorkTask(_StrictModel):
    id: StrictStr = Field(min_length=1, max_length=256)
    title: StrictStr = Field(min_length=1, max_length=256)
    state: Literal["pending", "in_progress", "completed", "failed", "cancelled"]
    detail: StrictStr | None = Field(default=None, max_length=16_000)
    startedAt: StrictStr | None = Field(default=None, min_length=1)
    completedAt: StrictStr | None = Field(default=None, min_length=1)


class GatewayWorkTool(_StrictModel):
    toolId: StrictStr = Field(min_length=1, max_length=256)
    name: StrictStr = Field(min_length=1, max_length=256)
    state: Literal["generating", "running", "completed", "failed"]
    args: Any | None = None
    result: Any | None = None
    summary: StrictStr | None = Field(default=None, max_length=16_000)
    progress: StrictStr | None = Field(default=None, max_length=16_000)
    durationMs: StrictInt | None = Field(default=None, ge=0)
    startedAt: StrictStr | None = Field(default=None, min_length=1)
    completedAt: StrictStr | None = Field(default=None, min_length=1)


class GatewayWorkApproval(_StrictModel):
    approvalId: StrictStr = Field(min_length=1, max_length=256)
    action: StrictStr = Field(min_length=1, max_length=256)
    reason: StrictStr = Field(min_length=1, max_length=16_000)
    args: Any | None = None
    createdAt: StrictStr = Field(min_length=1)
    expiresAt: StrictStr = Field(min_length=1)
    state: Literal["pending", "approved", "denied", "expired"]


class GatewayWorkArtifact(_StrictModel):
    artifactId: StrictStr = Field(min_length=1, max_length=256)
    name: StrictStr = Field(min_length=1, max_length=256)
    kind: StrictStr | None = Field(default=None, max_length=256)
    uri: StrictStr | None = Field(default=None, max_length=2_048)
    sizeBytes: StrictInt | None = Field(default=None, ge=0)
    summary: StrictStr | None = Field(default=None, max_length=16_000)
    createdAt: StrictStr | None = Field(default=None, min_length=1)


class GatewayWorkLog(_StrictModel):
    level: Literal["debug", "info", "warn", "error"]
    text: StrictStr = Field(min_length=1, max_length=16_000)


class GatewayWorkEvent(_StrictModel):
    type: GatewayWorkEventType
    runId: StrictStr = Field(min_length=1, max_length=256)
    conversationId: StrictStr = Field(min_length=1, max_length=256)
    eventId: StrictStr = Field(min_length=1, max_length=256)
    sequence: StrictInt = Field(ge=0)
    timestamp: StrictStr = Field(min_length=1)
    phase: StrictStr | None = Field(default=None, max_length=256)
    statusLabel: StrictStr | None = Field(default=None, max_length=256)
    statusDetail: StrictStr | None = Field(default=None, max_length=16_000)
    delta: StrictStr | None = Field(default=None, max_length=16_000)
    message: StrictStr | None = Field(default=None, max_length=16_000)
    summary: StrictStr | None = Field(default=None, max_length=16_000)
    tasks: list[GatewayWorkTask] | None = None
    tool: GatewayWorkTool | None = None
    approval: GatewayWorkApproval | None = None
    artifact: GatewayWorkArtifact | None = None
    progress: StrictStr | None = Field(default=None, max_length=16_000)
    log: GatewayWorkLog | None = None
    finalMarkdown: StrictStr | None = Field(default=None, max_length=32_000)
    stopReason: StrictStr | None = Field(default=None, max_length=256)
    errorCode: StrictStr | None = Field(default=None, max_length=256)
    errorMessage: StrictStr | None = Field(default=None, max_length=16_000)
    retryable: StrictBool | None = None

    @model_validator(mode="after")
    def validate_event_payload(self) -> "GatewayWorkEvent":
        common = {"type", "runId", "conversationId", "eventId", "sequence", "timestamp"}
        allowed_by_type = {
            "turn.started": {"phase", "statusLabel", "statusDetail"},
            "status.update": {"phase", "statusLabel", "statusDetail"},
            "assistant.interim": {"message"},
            "reasoning.delta": {"delta"},
            "plan.updated": {"tasks"},
            "tool.generating": {"tool"},
            "tool.started": {"tool"},
            "tool.progress": {"tool", "progress"},
            "tool.completed": {"tool"},
            "tool.output_risk": {"tool", "summary"},
            "approval.requested": {"approval"},
            "approval.resolved": {"approval"},
            "subagent.started": {"summary"},
            "subagent.progress": {"progress"},
            "subagent.completed": {"summary"},
            "background.completed": {"summary"},
            "artifact.created": {"artifact"},
            "log.line": {"log"},
            "assistant.delta": {"delta"},
            "turn.completed": {"finalMarkdown", "stopReason"},
            "error": {"errorCode", "errorMessage", "retryable"},
        }
        unexpected = self.model_fields_set - common - allowed_by_type.get(self.type, set())
        if unexpected:
            raise ValueError(f"event {self.type} contains fields for another event type: {sorted(unexpected)}")
        required = {
            "assistant.interim": ("message",),
            "reasoning.delta": ("delta",),
            "plan.updated": ("tasks",),
            "tool.generating": ("tool",),
            "tool.started": ("tool",),
            "tool.completed": ("tool",),
            "tool.output_risk": ("tool",),
            "approval.requested": ("approval",),
            "approval.resolved": ("approval",),
            "artifact.created": ("artifact",),
            "log.line": ("log",),
            "assistant.delta": ("delta",),
        }.get(self.type, ())
        if any(getattr(self, field) is None for field in required):
            raise ValueError(f"event {self.type} is missing its required payload")
        return self
